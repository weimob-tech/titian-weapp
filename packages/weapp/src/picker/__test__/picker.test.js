const simulate = require('miniprogram-simulate');
const { getComponent, getComponentId } = require('../../common/test/getComponent');
const { isIncomplete } = require('../const');
const getMock = require('./mock/index');

function getGroup(params) {
  const item = { ...params };
  const renderData = item.fn();
  if (!Array.isArray(renderData.options)) {
    renderData.options = [renderData.options];
    renderData.value = [renderData.value];
  }
  if (isIncomplete(renderData.options)) {
    renderData.options = [renderData.options];
    renderData.value = [renderData.value];
  }
  if (!Array.isArray(renderData.value)) {
    renderData.value = [];
  }
  renderData.value = renderData.value.filter(Boolean);

  if (!Array.isArray(item.result.value)) {
    item.result.value = [item.result.value];
  }
  if (!Array.isArray(item.result.options)) {
    item.result.options = [item.result.options];
  }
  item.result.value = item.result.value.map((i, index) => {
    if (typeof i !== 'object') {
      return {
        colAlias: `${index}`,
        value: i
      };
    }
    return i;
  });
  item.result.options = item.result.options.map((i, index) => {
    if (typeof i.colAlias === 'undefined') {
      return {
        colAlias: `${index}`,
        options: i
      };
    }
    return i;
  });
  renderData.options = renderData.options
    // eslint-disable-next-line @typescript-eslint/no-shadow
    .map((i, index) => {
      if (Array.isArray(i)) {
        return {
          colAlias: `${index}`,
          column: i
        };
      }
      if (typeof i.column !== 'undefined' && typeof i.colAlias !== 'undefined') {
        return i;
      }
      return undefined;
    })
    .filter(Boolean)
    .map((i, index) => ({
      ...i,
      _val: typeof renderData.value[index] === 'object' ? renderData.value[index].value : renderData.value[index]
    }));
  return { diffData: item, renderData };
}
describe('test component', () => {
  const pickerData = getMock()[1].list[0];
  it('should render success', () => {
    const comp = getComponent('picker');
    expect(comp).toMatchSnapshot();
  });

  it('定制 optionItemHeight string', async () => {
    const comp = getComponent('picker', {
      ...pickerData.fn(),
      optionItemHeight: '200rpx'
    });
    comp.triggerLifeTime('attached');
    comp.triggerLifeTime('ready');
    await simulate.sleep(0);
    expect(comp).toMatchSnapshot();
  });

  it('定制 optionItemHeight number ', async () => {
    const comp = getComponent('picker', {
      ...pickerData.fn(),
      optionItemHeight: 200
    });
    comp.triggerLifeTime('attached');
    comp.triggerLifeTime('ready');
    await simulate.sleep(0);
    expect(comp).toMatchSnapshot();
  });
  it('dispatch event ', async () => {
    const picker = getComponentId('picker');
    const pickerColumn = getComponentId('picker-column');

    const reachBottom = jest.fn();
    const reachTop = jest.fn();
    const toggleVisible = jest.fn();
    const onPickerChange = jest.fn();
    const { renderData } = getGroup(pickerData);
    const wrapper = simulate.render(
      simulate.load({
        usingComponents: {
          'ti-picker': picker,
          'ti-picker-column': pickerColumn
        },
        data: { ...renderData },
        template: `<ti-picker id="picker" visible-item-count="5" bind:change="onPickerChange"
                      bind:cancel="toggleVisible"  bind:confirm="toggleVisible" 
                      bind:reachBottom="reachBottom" bind:reachTop="reachTop">
                        <ti-picker-column 
                              label="{{ label }}" 
                              id="picker-column-{{idx}}"
                              wx:for-index="idx" 
                              value="{{item._val}}" 
                              wx:for="{{options}}" 
                              columns="{{ item.column }}" 
                              col-alias="{{item.colAlias}}"              
                              row-alias="{{ rowAlias }}"
                              use-row-index="{{ useRowIndex }}"
                        />
                    </ti-picker>
              `,
        methods: {
          onPickerChange,
          toggleVisible,
          reachBottom,
          reachTop
        }
      })
    );
    wrapper.attach(document.createElement('parent-picker'));
    const columnComp = wrapper.querySelector('#picker-column-0');
    const cancel = wrapper.querySelector('#picker >>> .titian-titlebar-left');
    cancel.dispatchEvent('tap');
    await simulate.sleep(0);
    expect(toggleVisible).toHaveBeenCalledTimes(1);
    toggleVisible.mockClear();
    columnComp.instance.synchronous({
      isTop: true,
      isBottom: false,
      rowIndex: 0,
      colAlias: columnComp.data.colAlias
    });
    await simulate.sleep(0);
    expect(reachTop).toHaveBeenCalledTimes(1);

    columnComp.instance.synchronous({
      isTop: false,
      isBottom: true,
      rowIndex: columnComp.data.columns.length - 1,
      colAlias: columnComp.data.colAlias
    });
    await simulate.sleep(0);

    expect(reachBottom).toHaveBeenCalledTimes(1);
  });
});
describe.each(getMock())('$value  $label', ({ list }) => {
  test.each(list)(` $label `, async (item) => {
    const picker = getComponentId('picker');
    const toggleVisible = jest.fn();
    const onPickerChange = jest.fn();
    const wrapper = simulate.render(
      simulate.load({
        usingComponents: {
          'ti-picker': picker
        },
        data: item.fn(),
        template: `<ti-picker
              id="picker"
              row-alias="{{ rowAlias }}"
              label="{{ label }}"
              value="{{ value }}"
              options="{{ options }}"
              cascade="{{ cascade || 'children' }}"
              use-row-index="{{ useRowIndex }}"
              visible-item-count="5"
              bind:change="onPickerChange"
              bind:cancel="toggleVisible"
              bind:confirm="toggleVisible" />`,
        methods: {
          onPickerChange,
          toggleVisible
        }
      })
    );
    wrapper.attach(document.createElement('parent-picker'));
    const confirm = wrapper.querySelector('#picker >>> .titian-titlebar-right');
    await simulate.sleep(0);
    expect(onPickerChange).toHaveBeenCalledTimes(1);
    expect(onPickerChange.mock.calls[0][0].detail).toStrictEqual(item.result);
    confirm.dispatchEvent('tap');
    await simulate.sleep(0);
    expect(toggleVisible).toHaveBeenCalledTimes(1);
  });
});

const group = getMock().filter((item) => item.value !== 'tree');

describe.each(group)('组合 $value  $label', ({ list }) => {
  test.each(list)(` $label `, async (item) => {
    const picker = getComponentId('picker');
    const pickerColumn = getComponentId('picker-column');

    const toggleVisible = jest.fn();
    const onPickerChange = jest.fn();
    const { diffData, renderData } = getGroup(item);
    const wrapper = simulate.render(
      simulate.load({
        usingComponents: {
          'ti-picker': picker,
          'ti-picker-column': pickerColumn
        },
        data: renderData,
        template: `<ti-picker
                      id="picker"
                      visible-item-count="5"
                      bind:change="onPickerChange"
                      bind:cancel="toggleVisible"
                      bind:confirm="toggleVisible" >
                        <ti-picker-column 
                              label="{{ label }}" 
                              wx:for-index="idx" 
                              value="{{item._val}}" 
                              wx:for="{{options}}" 
                              columns="{{ item.column }}" 
                              col-alias="{{item.colAlias}}"              
                              row-alias="{{ rowAlias }}"
                              use-row-index="{{ useRowIndex }}"
                        />
                      </ti-picker>
              `,
        methods: {
          onPickerChange,
          toggleVisible
        }
      })
    );
    wrapper.attach(document.createElement('parent-picker'));
    const confirm = wrapper.querySelector('#picker >>> .titian-titlebar-right');
    await simulate.sleep(0);
    expect(onPickerChange).toHaveBeenCalledTimes(renderData.options.length);
    renderData.options.forEach((i, index) => {
      const defaultValue = onPickerChange.mock.calls[index][0].detail;
      const val = [...diffData.result.value];
      const opt = [...diffData.result.options];
      val.length = index + 1;
      opt.length = index + 1;
      expect(defaultValue).toStrictEqual({ value: val, options: opt });
    });
    confirm.dispatchEvent('tap');
    await simulate.sleep(0);
    expect(toggleVisible).toHaveBeenCalledTimes(1);
    expect(toggleVisible.mock.calls[0][0].detail).toStrictEqual(diffData.result);
  });
});
