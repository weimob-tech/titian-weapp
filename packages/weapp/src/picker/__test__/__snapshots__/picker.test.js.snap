// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test component should render success 1`] = `
<main>
  <wx-view
    class="titian-picker"
  >
    <ti-popup-titlebar
      onlyCancel="{{true}}"
      catch:cancel="onCancel"
      catch:confirm="onConfirm"
    >
      <ti-title-bar
        extClass="ext-class"
        extLeftClass="ext-left-class titian-popup-titlebar-cancel"
        extRightClass="ext-right-class titian-popup-titlebar-confirm"
        extRightIconClass="ext-right-icon-class "
        extTitleClass="ext-title-class "
        bind:click="onClick"
      >
        <wx-view
          class="ext-class titian-titlebar"
          style=""
        >
          <wx-view
            class="ext-left-class titian-titlebar-left"
            bind:tap="onClickLeft"
          />
          <wx-view
            class="ext-title-class titian-titlebar-title"
          >
            <wx-view
              class="ext-main-title-class titian-titlebar-main-title"
              bind:tap="onClickTitle"
            >
              标题
            </wx-view>
          </wx-view>
          <wx-view
            class="ext-right-class titian-titlebar-right"
            bind:tap="onClickRight"
          />
        </wx-view>
      </ti-title-bar>
    </ti-popup-titlebar>
    <wx-view
      class="titian-picker-content ext-class"
      disableScroll="{{true}}"
      style=";"
    >
      <wx-view
        class="titian-picker-loading"
      >
        <ti-loading>
          <wx-view
            class="titian-loading-wrap"
            style=""
          >
            <wx-view
              class="ext-class titian-loading-circular titian-loading"
              style="padding:6rpx;width:60rpx;height:60rpx"
            >
              <wx-view
                class="titian-loading-view-circular titian-loading-view"
                style="width:60rpx;height:60rpx"
              >
                <wx-view
                  class="titian-loading-view-item"
                  style="width:60rpx;height:60rpx;border-width:5rpx"
                />
              </wx-view>
              <wx-view
                class="titian-loading-view-circular titian-loading-view"
                style="width:60rpx;height:60rpx"
              >
                <wx-view
                  class="titian-loading-view-item"
                  style="width:60rpx;height:60rpx;border-width:5rpx"
                />
              </wx-view>
              <wx-view
                class="titian-loading-view-circular titian-loading-view"
                style="width:60rpx;height:60rpx"
              >
                <wx-view
                  class="titian-loading-view-item"
                  style="width:60rpx;height:60rpx;border-width:5rpx"
                />
              </wx-view>
              <wx-view
                class="titian-loading-view-circular titian-loading-view"
                style="width:60rpx;height:60rpx"
              >
                <wx-view
                  class="titian-loading-view-item"
                  style="width:60rpx;height:60rpx;border-width:5rpx"
                />
              </wx-view>
              <wx-view
                class="titian-loading-view-circular titian-loading-view"
                style="width:60rpx;height:60rpx"
              >
                <wx-view
                  class="titian-loading-view-item"
                  style="width:60rpx;height:60rpx;border-width:5rpx"
                />
              </wx-view>
            </wx-view>
            <wx-view
              class="titian-loading-text"
              style=""
            />
          </wx-view>
        </ti-loading>
      </wx-view>
      <wx-view
        class="titian-picker-mask ext-mask-class"
      />
      <wx-view
        class="titian-picker-hairline ext-hairline-class"
        style=" height:0px;top : 0px"
      />
      <wx-view
        class="titian-picker-column-box"
      />
    </wx-view>
  </wx-view>
</main>
`;

exports[`test component 定制 optionItemHeight number  1`] = `
<main>
  <wx-view
    class="titian-picker"
  >
    <ti-popup-titlebar
      onlyCancel="{{true}}"
      catch:cancel="onCancel"
      catch:confirm="onConfirm"
    >
      <ti-title-bar
        extClass="ext-class"
        extLeftClass="ext-left-class titian-popup-titlebar-cancel"
        extRightClass="ext-right-class titian-popup-titlebar-confirm"
        extRightIconClass="ext-right-icon-class "
        extTitleClass="ext-title-class "
        bind:click="onClick"
      >
        <wx-view
          class="ext-class titian-titlebar"
          style=""
        >
          <wx-view
            class="ext-left-class titian-titlebar-left"
            bind:tap="onClickLeft"
          />
          <wx-view
            class="ext-title-class titian-titlebar-title"
          >
            <wx-view
              class="ext-main-title-class titian-titlebar-main-title"
              bind:tap="onClickTitle"
            >
              标题
            </wx-view>
          </wx-view>
          <wx-view
            class="ext-right-class titian-titlebar-right"
            bind:tap="onClickRight"
          />
        </wx-view>
      </ti-title-bar>
    </ti-popup-titlebar>
    <wx-view
      class="titian-picker-content ext-class"
      disableScroll="{{true}}"
      style=";height:330px"
    >
      <wx-view
        class="titian-picker-mask ext-mask-class"
      />
      <wx-view
        class="titian-picker-hairline ext-hairline-class"
        style=" height:110px;top : 110px"
      />
      <wx-view
        class="titian-picker-column-box"
      >
        <ti-picker-column
          data-col-index="{{0}}"
          extClass="ext-option-class"
          extOptionItemClass="ext-option-item-class"
          bind:reach-bottom="onReachBottom"
          bind:reach-top="onReachTop"
          bind:select="onSelect"
        >
          <wx-view
            $gdc="{{[Function]}}"
            $wxs:bind:touchend="{{[Function]}}"
            $wxs:bind:touchmove="{{[Function]}}"
            $wxs:bind:touchstart="{{[Function]}}"
            $wxs:change:y="{{[Function]}}"
            class="titian-picker-column-item-box ext-class"
            data-col-alias="{{0}}"
            data-height="{{110}}"
            data-row="{{0}}"
            data-size="{{5}}"
            data-sports="{{false}}"
            style=";top:110px"
            y="{{-0}}"
          >
            <wx-view
              $gdc="{{[Function]}}"
              $wxs:bind:tap="{{[Function]}}"
              $wxs:change:disabled="{{[Function]}}"
              class="titian-picker-column-item titian-picker-column-item-selected ext-option-item-class"
              data-col-alias="{{0}}"
              data-row="{{0}}"
              disabled=""
              style="height:110px"
            >
              
      选项一
    
            </wx-view>
            <wx-view
              $gdc="{{[Function]}}"
              $wxs:bind:tap="{{[Function]}}"
              $wxs:change:disabled="{{[Function]}}"
              class="titian-picker-column-item  ext-option-item-class"
              data-col-alias="{{0}}"
              data-row="{{1}}"
              disabled=""
              style="height:110px"
            >
              
      选项二
    
            </wx-view>
            <wx-view
              $gdc="{{[Function]}}"
              $wxs:bind:tap="{{[Function]}}"
              $wxs:change:disabled="{{[Function]}}"
              class="titian-picker-column-item  ext-option-item-class"
              data-col-alias="{{0}}"
              data-row="{{2}}"
              disabled=""
              style="height:110px"
            >
              
      选项三
    
            </wx-view>
            <wx-view
              $gdc="{{[Function]}}"
              $wxs:bind:tap="{{[Function]}}"
              $wxs:change:disabled="{{[Function]}}"
              class="titian-picker-column-item  ext-option-item-class"
              data-col-alias="{{0}}"
              data-row="{{3}}"
              disabled=""
              style="height:110px"
            >
              
      选项四
    
            </wx-view>
            <wx-view
              $gdc="{{[Function]}}"
              $wxs:bind:tap="{{[Function]}}"
              $wxs:change:disabled="{{[Function]}}"
              class="titian-picker-column-item  ext-option-item-class"
              data-col-alias="{{0}}"
              data-row="{{4}}"
              disabled=""
              style="height:110px"
            >
              
      选项五
    
            </wx-view>
          </wx-view>
        </ti-picker-column>
      </wx-view>
    </wx-view>
  </wx-view>
</main>
`;

exports[`test component 定制 optionItemHeight string 1`] = `
<main>
  <wx-view
    class="titian-picker"
  >
    <ti-popup-titlebar
      onlyCancel="{{true}}"
      catch:cancel="onCancel"
      catch:confirm="onConfirm"
    >
      <ti-title-bar
        extClass="ext-class"
        extLeftClass="ext-left-class titian-popup-titlebar-cancel"
        extRightClass="ext-right-class titian-popup-titlebar-confirm"
        extRightIconClass="ext-right-icon-class "
        extTitleClass="ext-title-class "
        bind:click="onClick"
      >
        <wx-view
          class="ext-class titian-titlebar"
          style=""
        >
          <wx-view
            class="ext-left-class titian-titlebar-left"
            bind:tap="onClickLeft"
          />
          <wx-view
            class="ext-title-class titian-titlebar-title"
          >
            <wx-view
              class="ext-main-title-class titian-titlebar-main-title"
              bind:tap="onClickTitle"
            >
              标题
            </wx-view>
          </wx-view>
          <wx-view
            class="ext-right-class titian-titlebar-right"
            bind:tap="onClickRight"
          />
        </wx-view>
      </ti-title-bar>
    </ti-popup-titlebar>
    <wx-view
      class="titian-picker-content ext-class"
      disableScroll="{{true}}"
      style=";height:0px"
    >
      <wx-view
        class="titian-picker-mask ext-mask-class"
      />
      <wx-view
        class="titian-picker-hairline ext-hairline-class"
        style=" height:0px;top : 0px"
      />
      <wx-view
        class="titian-picker-column-box"
      >
        <ti-picker-column
          data-col-index="{{0}}"
          extClass="ext-option-class"
          extOptionItemClass="ext-option-item-class"
          bind:reach-bottom="onReachBottom"
          bind:reach-top="onReachTop"
          bind:select="onSelect"
        >
          <wx-view
            $gdc="{{[Function]}}"
            $wxs:bind:touchend="{{[Function]}}"
            $wxs:bind:touchmove="{{[Function]}}"
            $wxs:bind:touchstart="{{[Function]}}"
            $wxs:change:y="{{[Function]}}"
            class="titian-picker-column-item-box ext-class"
            data-col-alias="{{0}}"
            data-height="{{0}}"
            data-row="{{0}}"
            data-size="{{5}}"
            data-sports="{{false}}"
            style=";top:0px"
            y="{{-0}}"
          >
            <wx-view
              $gdc="{{[Function]}}"
              $wxs:bind:tap="{{[Function]}}"
              $wxs:change:disabled="{{[Function]}}"
              class="titian-picker-column-item titian-picker-column-item-selected ext-option-item-class"
              data-col-alias="{{0}}"
              data-row="{{0}}"
              disabled=""
              style="height:0px"
            >
              
      选项一
    
            </wx-view>
            <wx-view
              $gdc="{{[Function]}}"
              $wxs:bind:tap="{{[Function]}}"
              $wxs:change:disabled="{{[Function]}}"
              class="titian-picker-column-item  ext-option-item-class"
              data-col-alias="{{0}}"
              data-row="{{1}}"
              disabled=""
              style="height:0px"
            >
              
      选项二
    
            </wx-view>
            <wx-view
              $gdc="{{[Function]}}"
              $wxs:bind:tap="{{[Function]}}"
              $wxs:change:disabled="{{[Function]}}"
              class="titian-picker-column-item  ext-option-item-class"
              data-col-alias="{{0}}"
              data-row="{{2}}"
              disabled=""
              style="height:0px"
            >
              
      选项三
    
            </wx-view>
            <wx-view
              $gdc="{{[Function]}}"
              $wxs:bind:tap="{{[Function]}}"
              $wxs:change:disabled="{{[Function]}}"
              class="titian-picker-column-item  ext-option-item-class"
              data-col-alias="{{0}}"
              data-row="{{3}}"
              disabled=""
              style="height:0px"
            >
              
      选项四
    
            </wx-view>
            <wx-view
              $gdc="{{[Function]}}"
              $wxs:bind:tap="{{[Function]}}"
              $wxs:change:disabled="{{[Function]}}"
              class="titian-picker-column-item  ext-option-item-class"
              data-col-alias="{{0}}"
              data-row="{{4}}"
              disabled=""
              style="height:0px"
            >
              
      选项五
    
            </wx-view>
          </wx-view>
        </ti-picker-column>
      </wx-view>
    </wx-view>
  </wx-view>
</main>
`;
